src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -8172,9 +8172,11 @@ StorageSnapshotPtr MergeTreeData::getStorageSnapshot(const StorageMetadataPtr &
 {
     auto snapshot_data = std::make_unique<SnapshotData>();
     ColumnsDescription object_columns_copy;
+    StreamSubscriptionPtr stream_subscription = nullptr;
 
     {
         auto lock = lockParts();
+        stream_subscription = subscribeForChanges();
```

Comment:
Generally storage will not have atomicity guarantees, but it can be implemented. It was just necessary to perform all operations with the subscription manager under the same lock as for the read/write pipelines.

For example StripeLog should create subscription in read method(because it has no snapshot realization) and with this subscription it starts pipeline, write on the other hand just push chunks to subscribers in the sink under the same lock. Serialization of {read, create subscription} = A and {write chunk to storage, push chunk to subscribers} = B will lead to one of executions: AB, BA. In first - read is already configured, so it will read from prepared snapshot(prepared in read) -> write to storage will not be visible, but will be visible subscribers chunk. In second case write will not see subscription for not started query -> chunk will be added to storage only.

SinkToSubscribers just naive implementation of this protocol compatible with any storage.

MergeTree does not have an atomic subscription at this stage of implementation. My idea was to change toSubscribersWrite for MergeTree - just do nothing in this method and push to subscribers chunks under the part-lock right in the MergeTreeData Transaction or under shared with it lock. 

This solution leads to 2 problems:
- write should always go to subscription manager, but it is a lightweight operation, because on write path only read lock is needed.
- write should create pulling pipeline from part and start reading from it - this can be a little expensive.

Reply:
> write should create pulling pipeline from part and start reading from it - this can be a little expensive.

To get the block(s) to push to subscribers? Can't we use the block(s) that was used to write the part?